### DISCUSSION TITLE
help me develop a low cost braille display

### DISCUSSION THREAD

**Discussion by discapacidad5** [parkrrrr/bdio#22] [General]
My language is Spanish I am using google translator so excuse me if it is not very clear

I saw this project, I found it interesting and very useful and I think you could help me with my project that seeks to get a braille display that can be used for up to less than 50 dollars

my big problem is how to build the braille HID for an esp32 device with the arduino ID or other possible alternative

You are now in aruino's HID library

an example from Espressif:
https://github.com/espressif/esp-idf/tree/master/examples/bluetooth/bluedroid/ble/ble_hid_device_demo
There is everything you need to get started based on a mouse / keyboard project.

Please try to compile and update the sample to ensure a successful development setup.
(introduction to ESP32: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/index.html)

And then start to change the HID descriptor to the example of your document.

but I don't know how to do that, could you help me?

my repository

https://github.com/brailletouch/Brailletouch

The encoder is practically the heart of the braille display. Most commercial braille displays have 40 or 80 braille cells. Since the most expensive on a braille display is braille cells, which cost more than $ 35 each, we are designing a new form of braille display with 40 virtual cells and a single physical braille cell, reducing most of the cost. . . . We are talking about a braille cell that can be made from 3D printed parts, I designed a different system. Instead of activating 40 cells at the same time and applying all the power to 40 physical cells, Brailletouch uses a physical encoder and a virtual braille display of 40 touch sensors placed in a matrix. In this way, the braille text is gradually displayed in a single braille cell as the virtual cells are touched. Parts can be easily printed.

My braille display design succeeds in reducing the cost of a braille display by over 90%. Currently, a braille screen has a cost between $ 1200 and $ 6000, being $ 1200 the smallest and most fragile, my proposal seeks to achieve a braille screen that can be manufactured between 100 and $ 150, it could even be much cheaper at the time from manufacturing to mass production. Basically, we could have a screen at an affordable price for all visually impaired and deafblind people.

My project is open source. The idea is to make available to anyone anywhere in the world a modifiable and improved code, files for 3D printers to print their parts and an assembly manual, so that anyone can download, print, assemble and use.

Its development is based on an esp32 microcontroller. (What is open source hardware)

**Comment by parkrrrr** [comment]
This isn't really an issue, and would probably be better posted in the discussion area, but I'll try to help. 

From your rendering, it appears that you are planning to have an 8-dot keyboard with three space keys, a d-pad and two buttons on the left, six buttons on the right, four buttons plus two pan buttons on the face, and some number of (virtualized) 8-dot Braille cells. 

Is that the only hardware configuration you expect to work with, or do you want the code to be easily adapted to different hardware designs? If you want code that adapts to different designs, you may find it worthwhile to modify some of my code to dynamically generate a report descriptor from some kind of data structure that describes the hardware. I can help you with that.

Beyond that, I have a spare ESP32 here, so let me see what I can do with the sample you linked to. 

I do want to ask - are you wanting this to be a Bluetooth device, or a USB device? I believe they both use the USB HID specifications at some level, but I'm not sure what other differences there are, as all of my experience is with USB HID.

**Reply by discapacidad5** [reply]
** From your rendering, it appears that you are planning to have an 8-dot keyboard with three space keys, a d-pad and two buttons on the left, six buttons on the right, four buttons plus two pan buttons on the face, and some number of (virtualized) 8-dot Braille cells. **

 It's a center space key and the Shift keys on the sides

There are 40 8 dot Braille cells (virtual with touch sensors) that I will later call up to display in a single physical cell.

in fact, it has the same number of keys as focus 40. plus a key combination equivalent to arrows on the navigation keyboard (left, right, up, down, enter, tab, uppercase + tab)

**Is that the only hardware configuration you expect to work with, or do you want the code to be easily adapted to different hardware designs? If you want code that adapts to different designs, you may find it worthwhile to modify some of my code to dynamically generate a report descriptor from some kind of data structure that describes the hardware. I can help you with that.**

Initially, if only that is the only hardware configuration I hope to work with, maybe in the future new devices based on this design can be created, but I think having a practical example is a good starting point to understand how it works. Braille HID and to be able to improve it and adapt it to others. in the future

**I do want to ask - are you wanting this to be a Bluetooth device, or a USB device? I believe they both use the USB HID specifications at some level, but I'm not sure what other differences there are, as all of my experience is with USB HID.**

Both USB and Bluetooth I hope I can take advantage of the ESP32's USB and Bluetooth interfaces

I have created a blank repository within my project for anyone who wants to help me. https://github.com/brailletouch/HID-braille-ESP32

Inside the readme file, put descriptive documentation and references to documents that I think can help a lot.

Thank you very much in advance for the support.
